### ISSUE TITLE
Some comments on the syntax

### ISSUE THREAD

**Issue by HenryLoenwind** [SerenityOS/jakt#1271]
Based on what is described in readme.md, which may be outdated. These are just random personal preferences/ideas.

```
    // Array<i64> with 256 elements, all initialized to 0.
    let values = [0; 256]
```

This looks like something that needs knowledge to read, i.e. you cannot guess what it means unless you know the language. My suggestion would be:

```
    let values = [0 x 256]
```

or

```
    let values = [(0) x 256]
```

and add

```
    let values = [(0, 1) x 128]
```

i.e. implement the "x" operator from Perl that creates repetitions of lists.

---

Arrays use []. Dictionaries use [], Sets use {}. I see no reason to clump arrays and dictionaries together like languages where they are the same data type. My suggestion would be to use [] for arrays, {} for dictionaries and <> for sets.

As a side effect, dictionaries could allow unquoted simple (a-zA-Z0-9_) keys in their declarations without needing a compiler that's guessing what it could mean.

---

With Sets a first-class data type, I would suggest adding language-level accessors instead of just using methods, e.g. ``` if 42 in magic_numbers``` and maybe even ``` if param_set & <42, 13, 4> // true if set contains all 3 numbers``` and ``` if param_set | <42, 13, 4> // true if set contains any of the 3 numbers``` or something like that (although this one needs some more thought!).

---

Quote operator. Perl has nifty quote operators, q and qq. Those allow you to specify which character you want to use as quote delimiter, with ' and " the defaults, respectively, in which case the operator can be left out. Example:

```Perl
$a = "\"\\\"";
$b = qq["\\"];
$c = q("\");
# $a === $b === $c
```

or

```Perl
$json = q!
  { "foo": {
    1, 2, 3,
  }, "bar": "something" ...
!;
```

Considering what a quote hell writing json or regexes in strings in other languages is, I suggest adopting a variant of that. It should be relatively straightforward to lex, especially if implemented with only a fixed subset for the allowed delimiter characters. (```$a = q🤢Hello🤢;``` is valid in Perl---that goes a bit far, doesn't it?)

One possibility would be to allow only ASCII quote characters (" ' ` ´) and have the difference between q and qq be that qq does everything " does, just with another delimiter and have q do no processing at all other than ending at the end delimiter.

---

> When calling a function, you must specify the name of each argument as you're passing it:

How about adding a third exception? Functions that only have one parameter. Unless you want to add polymorphism based on the name of the parameter(s) later, typing out the parameter name doesn't add anything. (Thinking about it, I may actually prefer the polymorphism. ```point.add(pixels: 10); point.add(pica: 10); point.add(cm: 1/2); ...```)

**Comment by HenryLoenwind** [comment]
I just commented the following on youtube, but it is a bit long. Not wanting to issue-spam I'm adding it here as comment.

---

I've been thinking about immutability a bit. I'd say object immutability and variable immutability are two orthogonal concepts.

That an object doesn't change its content doesn't have anything to do with a variable not being able to point to different objects. The number 5 is immutable; it's always five. But you can have a variable that's initialised with 5 and a constant that's defined as 5. The 5 doesn't care. And the other way round, neither the variable nor the constant care if whatever object they point to is immutable or not. Taken together, there can be:

* Mutable objects
* Immutable objects
* Mutable variables that contain mutable objects
* Immutable variables that contain mutable objects
* Mutable variables that contain immutable objects
* Immutable variables that contain immutable objects

Confusing? Indeed. That's why I'd suggest not merging the concepts. Let the programmer specify if the variable itself can change and if it can point to a mutable object independently.

However, "can point to a mutable object" needs further inspection, too. Does it mean, an object that is mutable can never be assigned? Or does it mean, no mutable operation can be done to the object when it is in that variable? To my understanding, "this object is mutable, but the following code says it won't do that to it" is a core concept of Jakt.

Also, assigning an immutable object to a variable that can contain mutable objects is a thing to think about. It seems to be a non-issue, like "I never declared I won't eat my car, why would I...it's inedible" (or in database terms: why duplicate the data?). Yet, for readability, it would make sense if that was clear when just looking at the consuming code without having to look up the class declaration. Also, the reality is that class declarations get changed over time. So a class that may be immutable in January, may become mutable in August.

The next thing is levels of mutability. What does it mean for an object to be immutable? Does it just mean that its fields cannot be changed? What if those unchangeable fields contain mutable objects? What if an unchangeable field contains an immutable object that represents the real-time clock? I'd classify as follows:

* An object is 1st class immutable if the values of its public members (i.e. public fields and public method return values) only depend on the identity of the object and (for methods) the given parameters.
* An object is 2nd class immutable if the values of its public members (i.e. public fields and public method return values) only depend on the identity of the object and (for methods) the given parameters and outside data.
* An object is 3rd class immutable if the values of its public members (i.e. public fields and public method return values) only depend on the identity of the object and (for methods) the given parameters and outside data and the values of contained data (i.e. mutable objects it holds in unchangeable fields).

Note that I anchored immutability on the public API of objects, not their side effects or internal storage. In many cases, this allows for objects that have side effects to be immutable. For example, an object that does vector transformations on 3D models can be immutable but still change data in the global profiler objects (i.e. record its runtime). Or allow methods to cache their return values (lazily compute on first use). I my eyes, this is more important than supporting 1st and 3rd class immutable objects.

By the way, it seems hard to have a compiler enforce this flavour of immutability, but compilers have come a long way in the past decades. Going through the parse tree and tagging variables with flags when they are assigned based on values used in the assignment and flags on the block the assignment happens in, as well as tagging blocks based on their conditional execution, is actually quite easy. And once that is in place, new flags can be introduced. For example, an "only for debug" flag that prevents the result of a method from accidentally depending on stuff that is flagged for debugging (who's never encountered a bug that only happens in production where debug code paths are disabled?).

(Reminder: Implicit data structures need a way of specifying immutability, too. An unchangeable variable with an immutable array of mutable objects and an unchangeable variable with a mutable array of immutable objects are somewhat different. In one case size() returns a value that's flagged "immutable", but in both cases, get() flags as "mutable".)

---

This isn't really an actionable thing, but more some thoughts on theory. I wouldn't presume I know Jakt well enough to attack its design, but I hope you who do know it well can gain from reading this.

**Comment by awesomekling** [comment]
Please don't spam the issue tracker with blog posts.
